crates/sui-replay/src/batch_replay.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::replay::LocalExec;
+use crate::types::ReplayEngineError;
+use futures::future::join_all;
+use parking_lot::Mutex;
+use std::collections::VecDeque;
+use std::sync::atomic::AtomicUsize;
+use std::sync::Arc;
+use sui_config::node::ExpensiveSafetyCheckConfig;
+use sui_types::base_types::TransactionDigest;
+use tokio::time::Instant;
+use tracing::{error, info};
+
+/// Given a list of transaction digests, replay them in parallel using `task_num` tasks.
+/// If `terminate_early` is true, the replay will terminate early if any transaction fails;
+/// otherwise it will try to finish all transactions.
+pub async fn batch_replay(
+    tx_digests: impl Iterator<Item = TransactionDigest>,
+    num_tasks: u64,
+    rpc_url: String,
+    expensive_safety_check_config: ExpensiveSafetyCheckConfig,
+    use_authority: bool,
+    terminate_early: bool,
+) {
+    let provider = Arc::new(TransactionDigestProvider::new(tx_digests));
+    let cancel = tokio_util::sync::CancellationToken::new();
+    let mut tasks = vec![];
+    let cur_time = Instant::now();
+    for _ in 0..num_tasks {
+        let provider = provider.clone();
+        let expensive_safety_check_config = expensive_safety_check_config.clone();
+        let rpc_url_ref = rpc_url.as_ref();
+        let cancel = cancel.clone();
```

Comment:
I think you can technically inline these now that they are not within a closure, right?